akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazyFlowSpec.scala

```diff
@@ -158,6 +158,17 @@ class LazyFlowSpec extends StreamSpec("""
       deferredMatVal.failed.futureValue shouldBe a[TE]
     }
 
+    "work for a single element when the future is completed after the fact" in assertAllStagesStopped {
+      val promise = Promise[Flow[Int, String, NotUsed]]()
+
+      val result: Future[immutable.Seq[String]] =
+        Source(List(1)).viaMat(Flow.lazyFutureFlow(() => promise.future))(Keep.right).runWith(Sink.seq)
+
+      promise.success(Flow[Int].map(_.toString))
```

Comment:
this races with data flow in the stream, you have no guarantee what happens first, is it the test completing the future or the LazyFlow consuming the first element.
I think u need more control here over the order of execution, either by introducing a second promise that the flow factory completes in order to signal the test code that the factory is invoked by the graph stage, or by other means (use a probe for source...).

Reply:
I'm happy to change this one, you're right in there being a race condition. Should I update the other existing 'after the fact' tests here too? They're all susceptible to this same race condition